### PR TITLE
Add suppport for gem commands in home (instead of only in the Ruby installation's bin dir)

### PR DESCRIPTION
**Pull request by paneidos**
This adds support for user-installed gems by looking at all directories in the gem path.

(Below is a clarification by @FooBarWidget)
`rbenv exec`, `rbenv rehash` etc. only look for gem commands in the Ruby installation's 'bin' directory. In other words, /usr/lib/rbenv/versions/xxx/bin/foo. Since a system-wide-installed Fullstaq Ruby installation is supposed to support per-user gems, it would be great if rbenv looks for gem commands in the user's home directory too (~/.gem/ruby/xxx/bin).

### CODE REVIEW

**Comment by FooBarWidget** [comment]
Hey, thanks for this PR and sorry for the lack of response! Apparently I didn't have notifications turned on for this repo. I'll have a look at this.

**Comment by paneidos** [comment]
> Hey, thanks for this PR and sorry for the lack of response! Apparently I didn't have notifications turned on for this repo. I'll have a look at this.

Hi, any chance of having this merged? At this point I've been running this branch in production for at least six months and even longer on my development machines.

**Comment by FooBarWidget** [comment]
@paneidos Yes, the feature looks acceptable. I'll start a code review.

File: rbenv.d/rehash/user-gems.bash [hunk 1]
```diff
@@ -0,0 +1,29 @@
+list_gem_dirs() {
```
Comment (FooBarWidget) [inline rbenv.d/rehash/user-gems.bash]: This function name does not seem to represent what the code actually does. It appears that the code prints command names instead of directory names.

File: rbenv.d/rehash/user-gems.bash [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+          for file in "${bindir}/"*; do
+            echo -n "${file##*/} "
+          done
+        done
+      fi
+    done
+  done
+}
+
+binstubs="`list_gem_dirs | sort -u`"
```
Comment (FooBarWidget) [inline rbenv.d/rehash/user-gems.bash]: I don't think this works. The function prints words, separated by spaces. But `sort -u` operates on a per-line basis.

File: rbenv.d/rehash/user-gems.bash (hunk 1 above)
Comment (FooBarWidget) [inline rbenv.d/rehash/user-gems.bash]: I think this file should have a comment at the beginning which summarizes what the code does.

File: libexec/rbenv-which [hunk 3]
```diff
@@ -60,7 +60,8 @@ for RBENV_COMMAND_PATH in "${RBENV_COMMAND_PATHS[@]}"; do
   fi
 done
 
-if [ "$RBENV_VERSION" != "system" ] && [ ! -d "${RBENV_ROOT}/versions/${RBENV_VERSION}" ]; then
+if [ "$RBENV_VERSION" != "system" ] && [ ! -d "${RBENV_ROOT}/versions/${RBENV_VERSION}" ] && \
```
Comment (FooBarWidget) [inline libexec/rbenv-which]: Thanks for finding this bug! I've split this bug fix into a separate commit, which I've already merged to master. Please rebase/merge.

File: rbenv.d/which/user-gems.bash [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      if [ -x "$path" ]; then
+        RBENV_COMMAND_PATHS+=("$path")
+      fi
+    done
+  fi
+else
+  gem="${RBENV_ROOT}/versions/${RBENV_VERSION}/bin/gem"
+  if [ -x "$gem" ]; then
+    OLDIFS="$IFS"
+    IFS=':' gempaths=(`$gem environment gempath`)
```
Comment (FooBarWidget) [inline rbenv.d/which/user-gems.bash]: Calling `gem environment gempath` is relatively slow, which makes 'rbenv which' much slower compared to before. On my Mac, `gem environment gempath` takes 0.5s. Its behavior is not that complex. Would it be a good idea to reimplement it ourselves?

File: libexec/rbenv-whence [hunk 5]
```diff
@@ -24,6 +24,19 @@ whence() {
     path="$(rbenv-prefix "$version")/bin/${command}"
     if [ -x "$path" ]; then
       [ "$print_paths" ] && echo "$path" || echo "$version"
+    else
+      gem="$(rbenv-prefix "$version")/bin/gem"
+      if [ -x "$gem" ]; then
+        OLDIFS="$IFS"
+        IFS=':' gempaths=(`$gem environment gempath`)
```
Comment (FooBarWidget) [inline libexec/rbenv-whence]: I think '$gem' here should be quoted.

There are a few more occurances of this.

**Review by FooBarWidget — CHANGES_REQUESTED** [review]

File: rbenv.d/which/user-gems.bash (hunk 4 above)
Comment (paneidos) [reply rbenv.d/which/user-gems.bash]: The path contains some properties found in `RbConfig`, so we unfortunately need to invoke ruby. While looking for an alternative I also noticed the method of appending `/bin` to the path is not reliable. I've looked at several methods in the past weeks and now settled on `Gem.path.map(&Gem.method(:bindir))` to get a correct list on multiple platforms. I'll push my new code soon.

File: rbenv.d/rehash/user-gems.bash (hunk 1 above)
Comment (paneidos) [reply rbenv.d/rehash/user-gems.bash]: I changed the name to be more descriptive.

File: rbenv.d/rehash/user-gems.bash (hunk 2 above)
Comment (paneidos) [reply rbenv.d/rehash/user-gems.bash]: Good catch, fixed it.

File: rbenv.d/rehash/user-gems.bash (hunk 1 above)
Comment (paneidos) [reply rbenv.d/rehash/user-gems.bash]: I added comments to the start of the file describing what is does.

File: libexec/rbenv-whence (hunk 5 above)
Comment (paneidos) [reply libexec/rbenv-whence]: All occurrences of this are now replaced.

File: libexec/rbenv-which (hunk 3 above)
Comment (paneidos) [reply libexec/rbenv-which]: I rebased my branch.

File: rbenv.d/which/user-gems.bash (hunk 4 above)
Comment (paneidos) [reply rbenv.d/which/user-gems.bash]: I changed it to use the new implementation, which seems to be faster.

Comment (FooBarWidget) [reply rbenv.d/which/user-gems.bash]: Cool, this is much faster!

**Review by FooBarWidget — APPROVED** [review]

**Comment by FooBarWidget** [comment]
Thanks for the changes! I'll release this soon.
